lib/Dialect/Calyx/Export/CalyxEmitter.cpp

```diff
@@ -801,6 +805,58 @@ void Emitter::emitSeqMemory(SeqMemoryOp memory) {
   os << RParen() << semicolonEndL();
 }
 
+void Emitter::emitInvoke(InvokeOp invoke) {
+  StringRef callee = invoke.getCallee();
+  indent() << "invoke " << callee;
+  ArrayAttr portNames = invoke.getPortNames();
+  ArrayAttr inputNames = invoke.getInputNames();
+  /// Because the ports of all components of calyx.invoke are inside a (),
+  /// here the input and output ports are divided, inputs and outputs store
+  /// the connections for a subset of input and output ports of the instance.
+  llvm::StringMap<StringRef> inputsMap;
+  llvm::StringMap<StringRef> outputsMap;
+  for (auto [portNameAttr, inputNameAttr, input] :
+       llvm::zip(portNames, inputNames, invoke.getInputs())) {
+    StringRef portName = cast<StringAttr>(portNameAttr).getValue();
+    StringRef inputName = cast<StringAttr>(inputNameAttr).getValue();
+    /// Classify the connection of ports,here's an example. calyx.invoke
+    /// @r(%r.in = %id.out, %out = %r.out) -> (i32, i32) %r.in = %id.out will be
+    /// stored in inputs, because %.r.in is the input port of the component, and
+    /// %out = %r.out will be stored in outputs, because %r.out is the output
+    /// port of the component, which is a bit different from calyx's native
+    /// compiler. Later on, the classified connection relations are outputted
+    /// uniformly and converted to calyx's native compiler format.
+    StringRef inputMapKey = portName.drop_front(2 + callee.size());
+    if (portName.substr(1, callee.size()) == callee) {
+      // If the input to the port is a number.
+      if (isa_and_nonnull<hw::ConstantOp>(input.getDefiningOp())) {
+        hw::ConstantOp constant = cast<hw::ConstantOp>(input.getDefiningOp());
+        APInt value = constant.getValue();
+        std::string mapValue = std::to_string(value.getBitWidth()) +
+                               apostrophe().data() + "d" +
+                               std::to_string(value.getZExtValue());
+        inputsMap[inputMapKey] = mapValue;
```

Comment:
I think the Windows failure is stemming from this. The PR is creating a `StringRef` to a `std::string` which is going to go out of scope. An ASAN build or running with valgrind would likely flag this, too.

Reply:
Thank you!Thank you, sir. I'll verify this.

Reply:
( ASAN report: https://gist.github.com/dtzSiFive/b2851d9be2a81b2aaee5a2851e7ddf27 )

Reply:
Good catch! It is surprising there is no ASAN build before merge.